### ISSUE TITLE
Positional but optional arguments

### ISSUE THREAD

**Issue by anntzer** [evanunderscore/defopt#36]
It sometimes makes sense for an argument to be optional but still specified by position, rather than by an explicit --option (e.g. the second argument to `ln`).  In argparse parlance, this is `nargs="?"`.

A reasonable way to specify such arguments could be e.g.
```
@defopt.run
def main(x, y=None, *, z=...): ...
```
i.e. POSITIONAL_OR_KEYWORD arguments that have a default are treated as `nargs="?"` arguments if the function also defines KEYWORD_ONLY arguments (which would still be treated as --options).  If you don't need any --option, you could always use a dummy argument, e.g.
```
@defopt.run
def main(x, y=None, *, _): ...
```
(once #34 is fixed, the dummy argument would disappear).

A more radical approach would be to say that *only* KEYWORD_ONLY arguments are --options, and everything else is positional, but that's probably not viable as long as you want to support Py2 :-)  (Yes, the current proposal does not allow creating `nargs="?"` arguments under Py2 either, but what can you do...)

**Comment by evanunderscore** [comment]
I don't think I like the idea of changing the behavior based on whether there are or aren't keyword-only arguments, since that may end up being somewhat surprising. Perhaps instead it could be an option to `run` that toggles how things are handled? That still doesn't help Python 2, but I'm not worried about it missing features.

**Comment by anntzer** [comment]
I don't feel very strongly about it (my preference, which is the last option, is unrealistic as long as you want to support py2), let me know how you want to name that kwarg to run.

**Comment by evanunderscore** [comment]
I intend to retain Python 2 support until its official EOL date, whenever that happens to be. However, I wouldn't be opposed to (sometime between now and then) changing the default for this proposed kwarg such that Python 2 users would have to opt back in to the old behavior.

**Comment by evanunderscore** [comment]
Regarding the naming of the argument, I'll have to think about it.

**Comment by evanunderscore** [comment]
The best thing I've come up with so far is `positional_kwargs`, but I'm happy to take better suggestions.

**Comment by anntzer** [comment]
Would the following behavior suit you?  (possibly `strict_kwonly`)
```
    :param bool use_kwonly: If `True` (the default), only keyword-only
        arguments are converted into command line flags; non-keyword-only
        arguments become optional positional command line parameters.  Note
        that on Python 2, setting this to False is the only way to obtain
        command line flags.
```
(I'm not sure having some extra deprecation period with a later switch for Py2 is going to make the transition any smoother, TBH).

**Comment by evanunderscore** [comment]
That sounds fine.

**Comment by evanunderscore** [comment]
Implemented in #41
